openmc/geometry.py

```diff
@@ -98,6 +100,191 @@ def export_to_xml(self, path='geometry.xml'):
+            generate it from the materials.xml file.
+
+        Returns
+        -------
+        openmc.Geometry
+            Geometry object
+
+        """
+        # Helper function to get value from an attribute/element
+        def get(elem, name, default=None):
+            if name in elem.attrib:
+                return elem.get(name, default)
+            else:
+                child = elem.find(name)
+                return child.text if child is not None else default
+
+        # Helper function for keeping a cache of Universe instances
+        universes = {}
+        def get_universe(univ_id):
+            if univ_id not in universes:
+                univ = openmc.Universe(univ_id)
+                universes[univ_id] = univ
+            return universes[univ_id]
+
+        tree = ET.parse(path)
+        root = tree.getroot()
+
+        # Get surfaces
+        surfaces = {}
+        periodic = {}
+        for surface in root.findall('surface'):
+            s = openmc.Surface.from_xml_element(surface)
+            surfaces[s.id] = s
+
+            # Check for periodic surface
+            other_id = get(surface, 'periodic_surface_id')
+            if other_id is not None:
+                periodic[s.id] = int(other_id)
+
+        # Apply periodic surfaces
+        for s1, s2 in periodic.items():
+            surfaces[s1].periodic_surface = surfaces[s2]
+
+        # Dictionary that maps each universe to a list of cells/lattices that
+        # contain it (needed to determine which universe is the root)
+        child_of = defaultdict(list)
+
+        for elem in root.findall('lattice'):
+            lat_id = int(get(elem, 'id'))
+            name = get(elem, 'name')
+            lat = openmc.RectLattice(lat_id, name)
+            lat.lower_left = [float(i) for i in get(elem, 'lower_left').split()]
+            lat.pitch = [float(i) for i in get(elem, 'pitch').split()]
+            outer = get(elem, 'outer')
+            if outer is not None:
+                lat.outer = get_universe(int(outer))
+            universes[lat_id] = lat
+
+            # Get array of universes
+            dimension = get(elem, 'dimension').split()
+            shape = np.array(dimension, dtype=int)[::-1]
+            uarray = np.array([get_universe(int(i)) for i in
+                               get(elem, 'universes').split()])
+            for u in uarray:
+                child_of[u].append(lat)
+            uarray.shape = shape
+            lat.universes = uarray
+
+        for elem in root.findall('hex_lattice'):
+            lat_id = int(get(elem, 'id'))
+            name = get(elem, 'name')
+            lat = openmc.HexLattice(lat_id, name)
+            lat.center = [float(i) for i in get(elem, 'center').split()]
+            lat.pitch = [float(i) for i in get(elem, 'pitch').split()]
+            outer = get(elem, 'outer')
+            if outer is not None:
+                lat.outer = get_universe(int(outer))
```

Comment:
Need to add `lat.outer` to `child_of`.  See the lattice_hex regression test for an example where it makes a difference.